From 7770d9bf9150b3157fd8a2e40f3089fc6bde51b2 Mon Sep 17 00:00:00 2001
From: Bogdan Hamciuc <bogdan.hamciuc@nxp.com>
Date: Tue, 17 Dec 2019 17:25:20 +0200
Subject: [PATCH 042/269] s32g: Restore Isolation Enable bit in Ncore

Clear CAIUTC[IsolEn] bit that we have set during early BL2 boot.

Upstream-Status: Pending 

Signed-off-by: Bogdan Hamciuc <bogdan.hamciuc@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
Issue: ALB-4277
---
 plat/s32g/s32g274a_bl2_el3.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/plat/s32g/s32g274a_bl2_el3.c b/plat/s32g/s32g274a_bl2_el3.c
index 8f0e2fa50..f7d5ddf7b 100644
--- a/plat/s32g/s32g274a_bl2_el3.c
+++ b/plat/s32g/s32g274a_bl2_el3.c
@@ -8,6 +8,7 @@
 #include <common/bl_common.h>
 #include <common/desc_image_load.h>
 #include <drivers/console.h>
+#include <lib/mmio.h>
 #include "s32g_ncore.h"
 #include "s32g_pinctrl.h"
 #include "s32g_clocks.h"
@@ -63,9 +64,18 @@ struct bl_load_info *plat_get_bl_image_load_info(void)
 void bl2_el3_early_platform_setup(u_register_t arg0, u_register_t arg1,
 				  u_register_t arg2, u_register_t arg3)
 {
+	uint32_t caiutc;
+
 	s32g_plat_config_pinctrl();
 	s32g_plat_clock_init();
 
+	/* Restore (clear) the CAIUTC[IsolEn] bit for the primay cluster, which
+	 * we have manually set during early BL2 boot.
+	 */
+	caiutc = mmio_read_32(S32G_NCORE_CAIU0_BASE_ADDR + NCORE_CAIUTC_OFF);
+	caiutc &= ~NCORE_CAIUTC_ISOLEN_MASK;
+	mmio_write_32(S32G_NCORE_CAIU0_BASE_ADDR + NCORE_CAIUTC_OFF, caiutc);
+
 	ncore_init();
 	ncore_caiu_online(A53_CLUSTER0_CAIU);
 
-- 
2.17.1

